packages/filesystem/src/reactive.ts

```diff
@@ -0,0 +1,242 @@
+      getTypeMap.get(path)?.[1](undefined);
+      getTypeMap.delete(path);
+      [...readdirMap.keys()].forEach(item => {
+        if (item.startsWith(`${path}/`)) {
+          getTypeMap.get(item)?.[1](undefined);
+          getTypeMap.delete(item);
+        }
+      });
+      readdirMap.get(getParentDir(path))?.[1](
+        items =>
+          items.filter(item => item === getItemName(path)) as [] | [string, ...string[]],
+      );
+    },
+  };
+  return fs;
+};
+
+/** makes an asynchronous filesystem reactive */
+export const createAsyncFileSystem = (adapter: AsyncFileSystemAdapter): AsyncFileSystem => {
+  const getTypeMap: ResourceMap<ItemType> = new Map();
+  const readdirMap: ResourceMap<[] | [string, ...string[]]> = new Map();
+  const readFileMap: ResourceMap<string> = new Map();
+  const fs: AsyncFileSystem = {
+    getType: path => {
+      if (!getTypeMap.has(path)) {
+        getTypeMap.set(
+          path,
+          createResource(() => adapter.getType(path)),
+        );
+      } else {
+        getTypeMap.get(path)![1].refetch();
+      }
+      const [fileType] = getTypeMap.get(path)!;
+      return fileType;
+    },
+    readdir: path => {
+      if (!readdirMap.has(path)) {
+        readdirMap.set(
+          path,
+          createResource<[] | [string, ...string[]]>(() => adapter.readdir(path), {
+            initialValue: [],
+          }),
+        );
+      } else {
+        readdirMap.get(path)![1].refetch();
+      }
+      const [files] = readdirMap.get(path)!;
+      return files;
+    },
+    mkdir: path =>
+      adapter.mkdir(path).then(() => {
+        path.split("/").forEach((_, index, parts) => {
+          const subPath = parts.slice(0, index + 1).join("/");
+          if (!getTypeMap.has(subPath)) {
+            fs.getType(subPath);
+          } else {
+            getTypeMap.get(subPath)![1].refetch();
+          }
+        });
+        readdirMap.get(getParentDir(path))?.[1].refetch();
+      }),
+    readFile: path => {
+      if (!readFileMap.has(path)) {
+        readFileMap.set(
+          path,
+          createResource(() => adapter.readFile(path), { initialValue: "" }),
+        );
+      } else {
+        readFileMap.get(path)![1].refetch();
+      }
+      const [data] = readFileMap.get(path)!;
+      return data;
+    },
+    writeFile: (path, data) =>
+      adapter.writeFile(path, data).then(() => {
+        readFileMap.get(path)?.[1].mutate(data);
+        const name = getItemName(path);
+        readdirMap
+          .get(getParentDir(path))?.[1]
+          .mutate(items =>
+            items.includes(name) ? items : ([...items, name] as [string, ...string[]]),
+          );
+      }),
+    rename: async (previous, next) => {
+      const previousType = await adapter.getType(previous);
+      const nextType = await adapter.getType(next);
+      if (previousType === "dir" && nextType === "file") {
+        throw new Error(`Cannot overwrite file "${next}" with directory "${previous}"`);
```

Comment:
I don't think hiding errors from the user is a good idea in that case. On the contrary, I believe useful error messages are a worthwhile feature even in production, because this is something that may fail there as well and you need to be aware to recover gracefully.